Add some enhancements to `jsonutils.nim`

bobeff opened the pull request:
* Use `jsonutils.nim` hookable API to add a possibility to deserialize JSON arrays directly to `HashSet` and `OrderedSet` types and respectively to serialize those types to JSON arrays.

* Also add a possibility to deserialize JSON `null` objects to Nim option objects and respectively to serialize Nim option object to JSON object if `isSome` or to JSON `null` object if `isNone`.
    
* Move serialization/deserialization functionality for `Table` and `OrderedTable` types from `jsonutils.nim` to `tables.nim` via the hookable API.
    
* Add object `jsonutils.Joptions` and parameter from its type to `jsonutils.fromJson` procedure to control whether to allow deserializing JSON objects to Nim objects when the JSON has some extra or missing keys.
    
* Add unit tests for the added functionalities to `tjsonutils.nim`.

bobeff committed:
Add some enhancements to `jsonutils.nim`

* Use `jsonutils.nim` hookable API to add a possibility to deserialize
  JSON arrays directly to `HashSet` and `OrderedSet` types and
  respectively to serialize those types to JSON arrays.

* Also add a possibility to deserialize JSON `null` objects to Nim
  option objects and respectively to serialize Nim option object to JSON
  object if some or to JSON `null` object if none.

* Move serialization/deserialization functionality for `Table` and
  `OrderedTable` types from `jsonutils.nim` to `tables.nim` via the
  hookable API.

* Add object `jsonutils.Joptions` and parameter from its type to
  `jsonutils.fromJson` procedure to control whether to allow
  deserializing JSON objects to Nim objects when the JSON has some
  extra or missing keys.

* Add unit tests for the added functionalities to `tjsonutils.nim`.

timotheecour commented on lib/pure/collections/sets.nim:
```diff
@@ -907,7 +909,52 @@ iterator pairs*[A](s: OrderedSet[A]): tuple[a: int, b: A] =
+    let foo = Foo(
+      hs: ["hash", "set"].toHashSet,
+      os: ["ordered", "set"].toOrderedSet)
+    assert $toJson(foo) == """{"hs":["set","hash"],"os":["ordered","set"]}"""
```
hash is subject to change, this could be tested in a hash-independent way for HashSet (at the expense of making test a bit more complex)

bobeff replied:
Here I made it simpler with only one element in the set and I added the more complex test in `tjsonutils.nim`.

timotheecour commented on lib/pure/collections/sets.nim:
```diff
@@ -907,7 +909,52 @@ iterator pairs*[A](s: OrderedSet[A]): tuple[a: int, b: A] =
+  ## 
+  ## See also:
+  ## * `toJsonHook proc<#toJsonHook,SomeSet[A]>`_
+  runnableExamples:
```
please also update tests/stdlib/tjsonutils.nim (with HashSet, OrderedSet, and if not already present, Table and OrderedTable)

since this is not meant to be used directly, the runnableExamples is less important than the test in tjsonutils

bobeff replied:
Done.

timotheecour commented:
great, that was on my roadmap, thanks for tackling this.
for `allowMissingFields` I'd prefer to wrap it into a single (optional) ref object (eg: `Joptions`) for following reasons:
* I expect more options will be needed for customization
* a single ref object is easier to forward to procs rather than passing each option individually; that is true for both internal procs (eg fromJson calling fromJsonFields) but also for user code.
* a set is not flexible enough as it only allows binary flags

so I'm suggesting this API, which is easier to extend:
```nim
type Joptions* = ref object
  allowMissingFields*: bool
proc initJoptions(): Joptions = ...
proc fromJson*[T](a: var T, b: JsonNode, opt = initJoptions()) = ...
```

EDIT: see rest of discussions, this is what we ended up with (object instead of ref object):
```nim
type Joptions* = object
  allowMissingFields*: bool
proc initJoptions(): Joptions = ...
proc fromJson*[T](a: var T, b: JsonNode, opt = initJoptions()) = ...
```

Araq commented:
Why not at least a ``var object`` instead? You don't take ownership of this thing.

timotheecour commented:
making it `object` instead of `ref object` is ok, but why make it `var`? it would prevent passing the options as an rvalue (and fromJson/toJson is not modifying joptions).

here's the modified API with object instead of ref object (but no `var`)
```nim
type Joptions* = object
  allowMissingFields*: bool
proc initJoptions(): Joptions = ...
proc fromJson*[T](a: var T, b: JsonNode, opt = initJoptions()) = ...
```

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +90,27 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
+    checkJson b.len == num, $(b.len, num, $T, b)
 
-proc fromJson*[T](a: var T, b: JsonNode) =
+proc fromJson*[T](a: var T, b: JsonNode, allowMissingFields = false) =
```
add a test in tjsonutils with and without `allowMissingFields`

bobeff replied:
Done.

Araq commented:
> here's the modified API with object instead of ref object (but no var)

Even better then.

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +90,27 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
+      elif not allowMissingFields:
         checkJson false, $($T, key, b)
-  checkJson b.len == num, $(b.len, num, $T, b) # could customize
+  if not allowMissingFields:
```
* need to document that `allowMissingFields` only refers to non-discriminant keys (and discriminant keys are still required for object variants, unless you find a good way to deal with missing discriminants, but this could be handled in future work)

* this not really correct now that you've introduced `allowMissingFields`;
we need to distinguish whether a key is in a but not in b and vice versa, and user should be able to allow either scenarios independently.

I suggest:
```nim
type Joptions = object
  allowMissingKeys: bool
  allowExtraKeys: bool
  # in future work: a key rename could be added
```
(keys makes it more obvious we're referring to the json key instead of the object field).

* the check `checkJson b.len == num, $(b.len, num, $T, b)` is now incorrect now that you're allowing `allowMissingKeys`; maybe something like:

```nim
for key, val in fieldPairs(a):
  ...
  if b.hasKey key:
    numMatched.inc
    ...

if not opt.allowExtraKeys:
  checkJson b.len <= numMatched, $(b.len, numMatched, $T, b)
```

and this should be exercised in tjsonutils

bobeff replied:
Here `numMatched` will be always smaller or equal than `b.len`. I think that the correct check is:
```nim
if opt.allowExtraKeys and opt.allowMissingKeys:
  discard
elif opt.allowExtraKeys:
  # This check is redundant because if here missing keys are not allowed, and
  # if `b.len < num` it will fail in the loop above but it is left for clarity.
  checkJson b.len >= num, $(b.len, num, $T, b)
elif opt.allowMissingKeys:
  checkJson b.len <= num, $(b.len, num, $T, b)
else:
  checkJson b.len == num, $(b.len, num, $T, b)
```

which is a little verbose and it can be written in shorter form as:
```nim
if not opt.allowExtraKeys and not opt.allowMissingKeys:
  checkJson b.len == num, $(b.len, num, $T, b)
elif not opt.allowExtraKeys and opt.allowMissingKeys:
  checkJson b.len <= num, $(b.len, num, $T, b)
```

but I preferred to write it in the first form because it seems clearer to me. I can change it if you prefer the second one.

bobeff replied:
I wrote `Joptions` object from the point of view of the JSON objects. Said with other words, extra keys and missing keys refer to JSON object keys. It can be also written to refer to Nim object fields. I'm not sure which is better and I could change it if you like?

timotheecour replied:
let's make sure it's correct first; right now I still don't think it's correct for the reasons I gave in https://github.com/nim-lang/Nim/pull/15133#discussion_r463775084 ; consider this:

```nim
type A = object
  a1,a2,a3: int
j = {"a3": 1, "a4": 2} # pseudo-code
```
with allowMissingKeys=true, and allowExtraKeys = false, your test passes but should fail because of extra key `a4`.
That's why you need `numMatched`

bobeff replied:
You can see the latest version of this in the last commit. You were completely correct about `numMatched`. If you don't find another not working test we still have to decide whether the version with only two `if`s isn't better?

timotheecour replied:
> I wrote Joptions object from the point of view of the JSON objects.

ya i think that's fine (hence "keys" and not fields). No strong opinion but i think it's clear.

>  we still have to decide whether the version with only two ifs isn't better?

agnostic to that but there is 1 more concern: code bloat. how about:

```nim
var ok = if ... elif .... else: ...
checkJson ok, $(json.len, num, numMatched, $T, json)
```
which reduces bloat.
note that code bloat can impact performance to some extent (IR cache)

bobeff replied:
Fixed.

timotheecour replied:
nice!

timotheecour commented:
one final note: if you prefer you can split your PR in 2, to move the `allowMissingFields` (embedded in an object) in a separate PR since the rest of this PR seems ready; up to you.

timotheecour commented on lib/pure/collections/tables.nim:
```diff
@@ -1750,9 +1750,56 @@ iterator mvalues*[A, B](t: var OrderedTable[A, B]): var B =
-
-
+type
+  SomeTable*[K, V] = Table[K, V] | OrderedTable[K, V]
```
does SomeTable need to be exported?
(if so, it raises questions regarding other tables like CountTable)

bobeff replied:
I exported it to be visible in the documentation that it is union type of `Table` and `OrderedTable`, because it is in the interface of `fromJsonHook` and `toJsonHook` procedures which must be exported. If it is not exported and not visible in the documentation it will be more confusing IMHO.

timotheecour replied:
yes, but it's not meant to be used directly, and exporting `SomeTable` leaves CountTable in an awkward place; how about this: get rid of `SomeTable` and inline its definition, eg:
```nim
proc fromJsonHook*[K, V, JN](t: var (Table[K, V] | OrderedTable[K, V]), jsonNode: JN) =
```
i tried, it works

bobeff replied:
Unfortunately, it doesn't work for documentation links.
```nim
##  * `toJsonHook proc<#toJsonHook,(Table[K,V]|OrderedTable[K,V])>`_
```
I tried several variants with parentheses and without, with `Table` only, with spaces around `|` and without.

Maye this is not very important to work correctly or I could remove `See also:` section entirely? With `SomeTable` typedef it still seems cleaner to me. What about if it has to be added also for other table types in the future. The signatures will become too long.

timotheecour replied:
you can simply use `toJsonHook` and hopefully docgen will eventually be smarter; it's ok to not have a hyperlink everywhere. The problem with export is it becomes part of the API, and harder to change; it'd be odd to have CountTable singled out.
> What about if it has to be added also for other table types in the future

we can revisit then ;-) (but the other way around is not possible without a breaking change)

bobeff replied:
Fixed.

timotheecour commented on tests/stdlib/tjsonutils.nim:
```diff
@@ -13,7 +13,16 @@ proc testRoundtrip[T](t: T, expected: string) =
   doAssert t2.toJson == j
 
-import tables
+template expect(Error: type, body: untyped) =
```
does `doAssertRaises` not work here?

bobeff replied:
I didn't know about the existence of `doAssertRaises`. I will fix it.

bobeff replied:
Done.

timotheecour replied:
(feel free to click "resolve conversation" when you resolve and there's no pushback)

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +102,36 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
   checkJsonImpl(cond, astToStr(cond), msg)
 
-template fromJsonFields(a, b, T, keys) =
+template fromJsonFields(a, b, T, keys, opt) =
```
optional nit: `s/b/j/` would make it clearer which is the json

bobeff replied:
I'm not sure what you mean? Do you want to change the signature to something like:
```nim
template fromJsonFields(obj, json, T, discKeys, opt)
```

timotheecour replied:
ya that would be even clearer

bobeff replied:
Done.

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +102,40 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
-template fromJsonFields(a, b, T, keys) =
-  checkJson b.kind == JObject, $(b.kind) # we could customize whether to allow JNull
+template fromJsonFields(obj, json, T, discKeys, opt) =
+  checkJson json.kind == JObject, $(json.kind) # we could customize whether to allow JNull
```
s/$(json.kind)/$json.kind/

bobeff replied:
Done.

timotheecour commented on lib/pure/collections/sets.nim:
```diff
@@ -907,7 +909,52 @@ iterator pairs*[A](s: OrderedSet[A]): tuple[a: int, b: A] =
+    let foo = Foo(
+      hs: ["hash"].toHashSet,
+      os: ["ordered", "set"].toOrderedSet)
+    assert $toJson(foo) == """{"hs":["hash"],"os":["ordered","set"]}"""
```
optional nit: you can make this test shorter (good for a runnableExamples) by using a tuple containing `hs`, `os` instead of introducing `Foo`
(likewise in the other runnableExample IIRC); eg:
`let foo = (hs: ..., os: ...)`

bobeff replied:
Fixed.

timotheecour commented on tests/stdlib/tjsonutils.nim:
```diff
@@ -119,5 +119,95 @@ template fn() =
+    fromJson(s, parseJson("""["one","two"]"""))
+    doAssert s == ["one", "two"].toHashSet
+    let jsonNode = toJson(s)
+    doAssert jsonNode.kind == JArray
```
can you make this shorter using sort? (and `import algorithm, sequtils`):
something like (untested):
```nim
let jsonNode = toJson(s)
# skip the other checks, they're redundant since it.str would fail if kind != JString, etc
doAssert jsonNode.elems.mapIt(it.str).sorted == @["one", "two"]
```

(short + correct  = good ;-) )

bobeff replied:
Fixed.

timotheecour commented on tests/stdlib/tjsonutils.nim:
```diff
@@ -119,5 +119,95 @@ template fn() =
+      doAssertRaises ValueError, fromJson(foo, json)
+      doAssertRaises ValueError,
+                     fromJson(foo, json, Joptions(allowMissingKeys: true))
+
```
sorry, 1 more complication.

* we need 1 more test: when object has an extra field set to a non-default value, it should not be reset to its default value; this allows one to update an existing object inplace via `fromJson` with only a subset of keys specified in json, keeping other fields intact (if user wants full reset, he can use `jsonTo`)

your tests don't exercise that. eg:
```nim
var guide = AboutLifeUniverseAndEverythingElse(answer = 15)
fromJson(guide, parseJson("""{"question":"foo"}"""), Joptions(allowMissingKeys: true))
doAssert guide == AboutLifeUniverseAndEverythingElse(answer: 15, question: "foo")
```

note: for non-variant objects it should be straightforward (hopefully); but for variant objects it may be a bit trickier; maybe the simplest is to check whether discriminant keys already match what's in json; if so, no need to reset the object.
how about:
```nim
a = initCaseObject(T, fun)
const keys = getDiscriminants(T)
fromJsonFields(a, b, T, keys, opt)

=>

const keys = getDiscriminants(T)
when keys.len == 0:
  # not a variant (no code bloat for the common case)
  fromJsonFields(a, b, T, keys, opt)
else:
  if not all_keys_match(keys): # pseudocode
    a = initCaseObject(T, fun)
  fromJsonFields(a, b, T, keys, opt)
```

* optional nit: consider using single asserts as above instead of 1 assert per field (shorter = better)

bobeff replied:
I fixed it and also I added the possibility for optional discriminant keys. You can check the code and the tests for the exact semantic. In short words, the currently set branch is taken when the discriminant key is missing from the JSON.

timotheecour replied:
I really want this PR to be merged (it's great work!) and the tests are good now, but something is bothering me with the latest commit: it introduces un-necessary copies with `var newObj: T; fromJsonFields(...); a = newObj` (and possibly extra complexity, but I'm not sure if smthg simpler is possible).

The whole point of `fromJson` was to allow inplace updates whereever possible. But now if even a single field is specified in json with `allowMissingKeys:true`, everything must be copied over still.

Let me think.

EDIT: ok how about the following, which makes the common case fast:
/cc @bobeff
```nim
const keys = getDiscriminants(T)
when keys.len == 0:
  # not a variant (`when` is important here, avoids cgen code bloat for the common case)
  fromJsonFields(a, b, T, keys, opt)
else:
  if discKeysMatch():
    # discriminant keys don't need to change, either bc no specified in json, or
    # specified in json but matches existing discriminant keys
    fromJsonFields(a, b, T, keys, opt)
  else:
    # edge case: discriminant keys require object reset, but we remember
    # the old fields in case these are not specified in json
    var newObj = initCaseObject(T, fun)
    fromJsonFields2(a, b, T, keys, opt, newObj)
    a = newObj
```

bobeff replied:
I introduced the copies intentionally, in order to achieve exception safety. Otherwise if in some moment exception is thrown by the procedure, the object can be left partially filled with old values and partially with the new ones. I many cases this can be completely fine but in some cases, it will be wrong. The choice was between performance and correctness.

But know it came to my mind that this can be better resolved on the caller side and I will follow your advice to avoid the copies when possible.

bobeff replied:
Done.

timotheecour replied:
> this can be better resolved on the caller side

indeed; so user can choose between no mutation under exception thrown (edge case) vs performance (common case)

bobeff commented:
I also implemented the serialization and deserialization of `Option[T]` types via the hookable API.

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +98,59 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
-template fromJsonFields(a, b, T, keys) =
-  checkJson b.kind == JObject, $(b.kind) # we could customize whether to allow JNull
+proc hasField[T](obj: T, field: string): bool =
+  for k, _ in fieldPairs(obj):
```
`for k in fields(obj)` seems more efficient?

bobeff replied:
Unfortunately, `fields` iterator yields objects values (only the second yield value of `fieldPairs`), but not the names of the fields (the first yield value of `fieldPairs`).

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +98,59 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
+  newDotExpr(obj, ident(name))
+
+template fromJsonFields(newObj, oldObj, json, discKeys, numMatched, opt) =
+  type T = typeOf(newObj)
```
`typeof`

bobeff replied:
Fixed.

timotheecour commented on lib/std/jsonutils.nim:
```diff
@@ -92,31 +98,59 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
+        fromJson(val, json[key])
+      elif opt.allowMissingKeys:
+        # if there is no discriminant keys the `oldObj` must always have the
+        # same keys and the new one. Otherwise we must check, because the could
```
nits: (can be fixed in subsequent cleanup PR)
s/if there is no discriminant keys/if there are no discriminant keys/
s/and the/as the/
s/because the could/because they could (?)

Yardanico replied:
@timotheecour you can use the GitHub suggest feature btw :)

bobeff replied:
Fixed.

timotheecour replied:
@Yardanico true true... on the fence

bobeff commented:
Hm. I just tested that my code is not working with nested variant objects.

bobeff commented:
@timotheecour But as it seems your initial code also does not work with nested variants. I propose if you have no other considerations to merge it as is, and to leave the support for nested variants as an open problem for future improvement.

timotheecour committed:
improve fromJsonFields

timotheecour commented:
@araq I'll let you take a look; it's unfortunately a bit more complex than I wish but most of the complexity comes from dealing with case objects + optional fields. Maybe there's something that could be done in the compiler to make this easier, but I'm not sure.

@bobeff 

* thanks for your patience with all my comments! excellent work, I think it's good to merge right now.
* you may want to update PR title + top-level msg to reflect the other things your PR does
* I've made the following improvement, which avoids copies in common case, feel free to incorporate it in your PR (I don't care about the commit, you can patch it manually or rebase+squash):
https://github.com/timotheecour/Nim/pull/338/commits/2582e49cd3ef5832a8ed6a0c7c013386610d5a4b
`https://github.com/timotheecour/Nim/pull/338/commits/2582e49cd3ef5832a8ed6a0c7c013386610d5a4b`
but can also be done as followup PR up to you
* in future PRs, it helps if you don't squash after review comments, so that reviewer can see the diff from the last time they reviewed ;-)
> Hm. I just tested that my code is not working with nested variant objects.

you could add a test case in tjsonutils under `when false:` or (if compiles but doesn't work at RT) with `if false:` + some comment; that failing test case can be done in a followup PR though

bobeff committed:
Add changelog entry for the jsonutils enhancements

bobeff committed:
Add TODO in `jsonutils.nim`

* Added an entry to "Future directions" section in `jsonutils.nim` as
  suggestion for future support of serialization and de-serialization of
  nested variant objects.

* Added currently disabled test case in `tjsonutils.nim` for testing
  serialization and de-serialization of nested variant objects.

bobeff commented:
@timotheecour @Araq  I updated the pull request title and top-level comment to incorporate the information for the other improvements, cherry-picked Timothee's commit with the additional optimization of the copies, added a comment in `Future directions:` section, as a suggestion for implementing support for serialization and de-serialization of nested variants and added a disabled test case for them. I also added a changelog entry for the current additions.

timotheecour commented:
@Araq if you don't have any further comments I'll go ahead and merge this soon

Araq commented:
Er. The idea was that the collections do not know about json, now tables.nim lacks the json import but clearly knows about json. (The ``auto`` return type is a figleaf.)

Btw I'm on holidays, so please excuse my rare responses.

bobeff commented:
@Araq As I see the things, either JSON module has to know about all collections or the collections have to know about the JSON. Is it possible to implement it fully decoupled?

Araq commented:
Previously jsonutils.nim managed to do that, correct me if I'm wrong.

bobeff commented:
@Araq But previously `jsonutils.nim` knew about the tables. If we add dependencies to all collections we want to serialize to it, we will be back to the state from which we want to escape with the hookable API. Other variant is to add separate files dedicated to serialization for each collection: `tablesserialization.nim`, `setsserialization.nim`, `optionsserialization.nim` and etc.

Araq commented:
> But previously jsonutils.nim knew about the tables.

That's fine and you can keep adding deps to jsonutils. Until concepts are ready for this, I hope.

timotheecour commented:
I don't understand the problem; this PR doesn't add any dependency (and it removes `import tables` dependency from `import jsonutils`; the fact that json already imports tables makes that last point less relevant for now but could become relevant). The important point is that it supports Option+collections/sets (de)serialization without either module depending on the other, so that user code can import both and (de)serialize data containing Option/sets.

> Previously jsonutils.nim managed to do that, correct me if I'm wrong.

this PR follows the same logic as what jsonutils did for `strtabs` (for the 1st half of this PR, the rest of this PR is addressing separate issue of handling missing/extra keys)

Araq commented:
Well tables.nim now contains 

```nim

proc toJsonHook*[K, V](t: (Table[K, V] | OrderedTable[K, V])): auto = ...

```

Just because that happens to compile without ``import json`` doesn't mean tables.nim does not depend on json.nim. In fact, it *only* compiles because of the "auto" return type. Collections should not be aware of json, json is just one data format among many. 

For a comparison, take a look at https://en.cppreference.com/w/cpp/container/unordered_map . No reference to json in sight, a much better design.

timotheecour commented:
> For a comparison, take a look at

not a fair comparison since it leaves json (de)serialization up to the user to implement

> That's fine and you can keep adding deps to jsonutils

that's a worse solution, I don't want to import "the world" when I import jsonutils, not to mention cyclic dependency issues. The current approach avoids cyclic dependency issues as well as "import the world", which is the most important point, but yes it does make tables/options/sets aware of json, but that's just 2 extra symbols per collection.

The only other practical way I'm aware of that avoids this is:
* what @bobeff suggested, which is moving fromJsonHook /toJsonHook pairs to separate modules: tablesserialization, setsserialization, optionsserialization, strtabsserialization. Although it should probably be instead: json_tables, json_sets, json_options, json_strtabs since there are other formats.
Since modules are cheap (https://github.com/nim-lang/Nim/pull/14614) that won't add compile time cost, but it does add a bit of complexity for a common case (if user forgets to import these, the default serialization will be used) 

* making fromJsonHook / toJsonHook format-agnostic, eg: deserializeHook/serializeHook, eg:
```nim
proc deserializeHook*[T, JN](self: var Option[T], j: JN, deserialize: alias, isNull: alias) =
  if isNull(j):
    self = some(deserialize(jsonNode, T))
  else:
    self = none[T]()

proc serializationHook*[T](self: Option[T], serialize: alias, newNull: alias): auto =
  if isSome(self):
    serialize(get(self))
  else:
    newNull()
```
this is IMO the cleanest way (avoids generic instantiation caching issues and strictly more flexible).

But AFAIK that requires https://github.com/nim-lang/Nim/pull/11992 (my `alias` PR) to pass un-instantiated generic symbol (eg `serialize`, `deserialize`) to a proc; and this approach could always be done in some future work once #11992 is merged (not loosing hope).

c-blake commented:
JSON/ASCII-like formats while popular are highly inefficient in both space and time. Where they are most inefficient is on large objects which collections are more likely to be.  So, -1 from me on hard-coding "JSON-ness" into all the collections/places most likely to want something else the moment it is available. Some kind of general purpose/format-agnostic serialization might be fine, but I'm not sure it requires the alias stuff.  Maybe @Araq can figure out something simpler when his holidays are over.  I do sort of feel like a module intended to serialize the type world and importing the type world are a natural pairing, though I can see why you want to avoid it.  Hmm.  Tagging @Vindaar since I know he works on HDF5 serialization/deserialization in Nim and might have useful input.

Araq commented:
> making fromJsonHook / toJsonHook format-agnostic, eg: deserializeHook/serializeHook,

Well collections expose iterators for this task like ``items`` and ``pairs``.

c-blake commented:
>not a fair comparison since it leaves json (de)serialization up to the user to implement

Ok, but this works now with `marshal` not importing `sets`...
```Nim
import sets, marshal
template testRoundTrip(x) =
  echo to[typeof(x)]($$x) == x
testRoundTrip ["one", "two"].toOrderedSet
```
So, maybe I don't understand the point of `jsonutils`.  From the module itself, it looks like simple things like wanting to reject some fields if present, inject other fields if missing, and rename other fields, why not just add some defaulted `inject, reject: HashSet[string]` and `rename: Table[string,string]` params to `marshal` module's `load` & `store`?  That would seem to preserve "loose coupling" properties, but maybe not provide total generality.

Anyway, is there a concrete use case that is driving all this?  Often that helps to scope API/dependency choices..So, I went looking at https://github.com/nim-lang/Nim/pull/14549 and https://github.com/nim-lang/Nim/pull/14563 and found references to creating RFCs to discuss this, which seems to not ever have happened.  It seems like it all started just from making `strtabs` more like regular `tables` and is now kinda spiraling into complexity.

Maybe before we go much further here, someone should be an RFC on this serializability stuff (I personally strongly prefer the shorter and less ambiguous verb "marshal".  A lot of times people use "serialize" to mean "synchronize access over time", not merely "lay out in space", but I have literally never seen "marshal" used for anything but a layout/format in a programming context.)

bobeff committed:
Move JSON hooks to `jsonutils.nim`

Move `fromJsonHook` and `toJsonHook` procedures for different types to
`jsonutils.nim` module to avoid a dependency of collections modules to
the `json.nim` module.

The hooks are removed from the following modules:

  * `tables.nim`

  * `sets.nim`

  * `options.nim`

  * `strtabs.nim`

bobeff commented:
@Araq @timotheecour I moved all the hooks from the collections modules, more specifically from
* `tables.nim`
* `sets.nim`
* `options.nim`
* `strtabs.nim`

Now the collections modules don't depend on JSON, but `jsonutils.nim` imports all collections for which a json serialization/de-serialization is implemented, as Araq suggests to be more acceptable solution for him.

When implementing the `runnableExamples` for `StringTableRef` hooks I found a problem with it does not have implemented comparison. It is only comparing the refs, but not the actual objects as maybe it is intended, but the default comparison of derefed objects doesn't work correct either, because it is not order-independent. The tests in the runnable examples work correct by chance,  but it has to be decided on how to fix the problem.

* The most obvious solution is to implement ```proc `==`*(lhs, rhs: StringTableObj)```, but all other public methods in the module are implemented for `StringTableRef` and not for `StringTableObj`.

* If it is implemented like ```proc `==`*(lhs, rhs: StringTableRef)``` this will possibly be breaking change for someone's code which currently relies on shallow pointer comparison.

Araq commented:
It's called a ``StringTableRef`` so comparison should follow the builtin reference semantics. For tests write a custom compare proc, do not add to the public interface just for testing purposes.

bobeff commented:
> For tests write a custom compare proc, do not add to the public interface just for testing purposes.

I already have such a procedure written
```nim
proc `==`*(lhs, rhs: StringTableRef): bool =
  ## Compares two objects pointed by `StringTableRef`s by using the more
  ## restrictive of the two comparison modes.
  if lhs.unsafeAddr == rhs.unsafeAddr:
    # if it is the same table
    return true
  if lhs.len != rhs.len:
    # if the number of elements is different
    return false
  # `rhs` must have the more restrictive mode
  let (lhs, rhs) = if lhs.mode < rhs.mode: (rhs, lhs) else: (lhs, rhs)
  for key, val in pairs(lhs):
    if not rhs.hasKey(key):
      # if some of the `lhs`'s keys is missing from `rhs`
      return false
    if not myCmp(rhs, rhs.getOrDefault(key), val):
      # if the some of the keys have a different corresponding value
      return false
  return true
```
but I think that it is not appropriate to add such long code in the `runnableExamples` section, and after the procedure is not public it cannot be seen in it otherwise. Even if I strip all the comments and the documentation it is still at least 12 lines long, which is twice the size of the current code.

After the current code with the dereference is working for the test data (although it is not generally correct) I propose to finally put the end of this prolonged issue, if there are no other considerations.

Araq commented:
Er, tests != runnableExamples, examples can remain not to require == and the fact that you "already" wrote == for StringTableRef misses the point completely, I do not want to override the reference based equality for a Ref type. One point out of many, that's a breaking change.

bobeff committed:
Add some tests about `StringTableRef`

Add tests for `StringTableRef`'s `fromJsonHook` and `toJsonHook` to
`tjsonutils.nim`.

bobeff commented:
> Er, tests != runnableExamples, examples can remain not to require ==

Well, the JSON serialization/de-serialization code for `StringTableRef` is not written by me and I only moved it from the `strtabs` module to the `jsonutils` module for uniformity with the approach taken for the other collections. I only added the `runnableExamples` section also for uniformity with other `fromJsonHook` and `toJsonHook` procedures all of which have such.

After that code is not new I expected it to already have tests, which I see now apparently is not the case. In the last commit I added tests about `StringTableRef`'s `fromJsonHook` and `toJsonHook` in `tjsonutils.nim` without the comparison procedure, because the test framework represented by the `testRoundtrip` procedure only compares JSON strings and not objects.

bobeff committed:
Disable a warning in `jsonutils.nim`

Mark `fun` template in `jsonutils` module with `{.used.}` pragma in
order to disable `[XDeclaredButNotUsed]` hint. The template is actually
used by the `initCaseObject` macro in the same module.

bobeff commented:
Hi,

Are there any other suggestions about this pull request? I fixed everything which I was told, but it is still in limbo ...

Clyybber reviewed (approved):
I think Joptions is ugly and too generic of a name (sorry for the back and forth, I know timothee advised you to use it), I think we should get rid of that unneccessary sophistication and just use optional proc params directly.
Otherwise LGTM :+1:

bobeff commented:
> I think we should get rid of that unnecessary sophistication and just use optional proc params directly.

I'm rather on Timothee's side. If we add other options in the future, they will become too many procedure parameters. As Timothee himself pointed out:
> I expect more options will be needed for customization

> a single ref object is easier to forward to procs rather than passing each option individually; that is true for both internal procs (eg fromJson calling fromJsonFields) but also for user code.

timotheecour commented:
The version I approved was before the commit `Move JSON hooks to jsonutils.nim` which IMO goes in the wrong direction; compared to right before that commit, `jsonutils` now adds a (recursive) dependency on at least 7 modules: `os`, `osseps`, `pathnorm`, `posix`, `sets`, `strtabs`, `times`, which have nothing to do with serialization.

In particular, `os`, `posix` and `times` are not lightweight dependencies. 

This negates the whole point of the json hooks I had introduced in jsonutils (https://github.com/nim-lang/Nim/pull/14563), and prevents using jsonFrom, jsonTo in those modules due to cyclic dependency issues. Having toJsonHook/fromHsonHook in collections is arguably better, it just adds 2 symbols to collections but doesn't cause extra dependencies or prevent use of jsonutils in those modules.

See also https://github.com/nim-lang/Nim/pull/15133#issuecomment-670957080 above and https://github.com/nim-lang/Nim/pull/14549#issuecomment-639377654, it looks like we're going in circles...
> json should be light and unrelated to any OS specific stuff, but is now a kitchen sink of dependencies; here's the total imports for import json: [...]

c-blake commented:
It sounds to me like @Araq & @timotheecour have conflicting visions of future directions here leading to said circularity.  So, again there should be a How-To-Marshal-in-Nim RFC to get more exposure and again concrete use cases would help clarify these conflicting theoretical concerns.

timotheecour commented:
I'm not against an RFC but I thought the conclusion had already been reached that serialization should not import "the world":

https://github.com/nim-lang/Nim/pull/14549#issuecomment-638724320

> @araq: Sure, my major gripe is a json.nim that imports all the stdlib collections, there should be a better way. At least it's better than everything depending on json though...

https://github.com/nim-lang/Nim/pull/14549#issuecomment-639990793

>> we have dead code elimination anyway, so I don't see it being that bad

> it still needs to be parsed and is not cheap, eg look at this: #14179
simply adding dependency on os (without using any symbols) yields a 30x slowdown in that case
and it can lead to cyclic dependency issues

Adding (recursive) dependency on os+times+the other modules (see above) IMO makes this PR non-mergeable; it was perfectly fine before the commit `Move JSON hooks to jsonutils.nim`

c-blake commented:
Well, I read https://github.com/nim-lang/Nim/pull/14549#issuecomment-639350606 as @PMunch and @Araq both also wanting an RFC about the broader issues.

Clyybber commented:
@timotheecour None of these modules import jsonutils, which recursive/cyclic dependencies are you talking about?
> In particular, os, posix and times are not lightweight dependencies.

`posix` and `times` are imported *by* `os` which is imported by `strtabs` (the fact that strtabs imports `os` is more of a strtabs "issue" rather than an issue that concerns this PR).
Nothing in this PR explicitly imports any of the modules you listed.

bobeff commented:
I think that it is not so important whether all collections will depend on JSON or there will be a module dedicated to collections serialization importing all collections or will be separate modules for JSON serialization for each collection. Everything I want is to have capabilities to serialize/de-serialize collections and to allow missing/extra keys. This functionality was already started in the way proposed by Timothee in `strtabs.nim`. I implemented it following his design and then changed it in the way suggested by Andreas.

Feel free to revert the e61da21 if you again decide the other way, but I find the current situation with two separate APIs
* one in `json.nim`
* another in `jsonutils.nim` which is extended in this pull request.

both of them not complete to be worse than not an ideal solution or even a completely missing solution.

timotheecour commented:
@Clyybber the problem is two-fold; after the commit `Move JSON hooks to jsonutils.nim`, you can't have `jsonTo`, `jsonFrom` without importing all those recursive dependencies (os, times etc), even if you don't care about strtabs or don't need to serialize a `StringTableRef` in your code; even simply parsing modules is not cheap, see https://github.com/nim-lang/Nim/issues/14179 which caused a 30x slowdown by adding os dependency and later got fixed. 

> None of these modules import jsonutils, which recursive/cyclic dependencies are you talking about?

that commit doesn't create a cyclic dependency (that wouldn't even be possible) but prevents future imports which would create cyclic dependencies.

these problems were nonexistant before this commit.

bobeff commented:
> For a comparison, take a look at https://en.cppreference.com/w/cpp/container/unordered_map . No reference to json in sight, a much better design.

@Araq, @timotheecour This is not a completely fair comparison because even today there is no notion for JSON in the entire C++ standard library. But from the other side, the most modern languages which have such support like [C#](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-how-to) for example have dedicated modules for serialization. Maybe simply the too generic name `jsonuitls.nim` have to be changed to something like `jsonserialization.nim`?

> if you don't care about strtabs or don't need to serialize a StringTableRef in your code; even simply parsing modules is not cheap

The other variant which could be acceptable for both of you as it was mentioned before is to have a separate module for JSON serialization for each collection type like `strtabsjsonserialization.nim`, `tablesjsonserialization.nim`, `setsjsonserialization.nim`, `optionsjsonserialization.nim` and so on? I think that this will be a nice compromise, between the two points of view.

timotheecour commented:
> The other variant which could be acceptable for both of you

as I mentioned earlier I would be totally ok with that; modules are cheap.

bobeff commented:
@Araq @timotheecour If Araq also approves it I will make it that way.

timotheecour commented:
how about:
strtabsjsonserialization.nim => strtabs_json.nim (etc) ? non ambiguous and shorter

bobeff commented:
@timotheecour Ok but let us wait for Andreas to approve the design and the names of the modules because I already feel a little bored from this issue and I hope this to be the last time I have to change the things.

zah commented:
FWIW, this approach with a separate "bridge" module (e.g. `strtabs_json`) is what I use in nim-json-serialization as well:
https://github.com/status-im/nim-json-serialization/blob/master/json_serialization/std/sets.nim

I've made it a bit easier for the users to import a single module by re-exporting the the bridged modules from there:

```nim
import json_serialization/std/tables

var t = initTable[string, int]()
```

c-blake commented:
As mentioned before, using a more general term ("marshal" since "serialization" is potentially confusing and different from the current "marshal" module: lose-lose) instead of "json" might be more forward looking..that is if this module hook design can support other marshalled formats (e.g. some 10+x faster binary one someday).  If it can only support json for whatever reasons then "json" is fine, though that seems like an unfortunately limited design for the stdlib.

bobeff commented:
@c-blake Except if we won't also have separate modules for separate marshaling formats e.g. `tables_json.nim`, `tables_xml.nim`, `tables_binary.nim` and so on. Here we have the same problem with the multiple imports and the compilation speed. If I want to marshal only to JSON or only to some binary format why I have to import the modules for all other formats?

c-blake commented:
Ok.  In that case I would recommend a subdirectory `json/` (somewhere..maybe `std/marshal/json` if that won't conflict with `std/marshal.nim`?) to be followed (someday, possibly) by `std/marshal/xml/` `std/marshal/binary/`, etc.  Sounds like there are at least 4 stdlib names under such dirs (`sets`, `tables`, `strtabs`, `options`) so far.  Maybe more could happen.

bobeff commented:
Ok. I find the structure proposed by c-blake to be appropriate, but I also think that the current generic functionality in `jsonutils.nim` also have to be moved somewhere in this structure under some name like `std/marshal/json/json.nim` for example and `jsonutils.nim` to be deprecated. It will be good the old serialization API in `json.nim` also to be deprecated to not confuse the people what to use.

c-blake commented:
Co-locating the `jsonutils`-equivalent near the fork for type-special-purpose modules sounds reasonable to me.  Let's see if @Araq likes this new direction.  Don't know about deprecating `json.nim` or what to do about the `marshal.nim` module in light of all this.  Those also seem like questions for @Araq.

bobeff commented:
> Don't know about deprecating json.nim

To be clear I don't mean deprecation of the entire `json.nim`, but only the part related to serialization/de-serialization of different standard data types. The part dedicated to working with `JsonNode`s should stay as is.

Araq commented:
I'm writing an RFC that describes in detail how this feature should be designed. Once we agreed on the RFC, we can implement it and deprecate what doesn't fit.

Araq commented:
RFC is here, https://github.com/nim-lang/RFCs/issues/247

bobeff commented:
As this is a working backward compatible solution and currently I don't have time to reimplement it according to the RFC,  is there any chance to merge this pull request as a temporary solution until the RFC is implemented?

Araq commented:
Ok, will review it once again.
